server/src/compression.rs

```diff
@@ -0,0 +1,473 @@
+use crate::models;
+use crate::models::{gorilla::Gorilla, pmcmean::PMCMean, swing::Swing, ErrorBound, SelectedModel};
+use crate::storage::StorageEngine;
+use crate::types::{Timestamp, TimestampArray, TimestampBuilder, Value, ValueArray, ValueBuilder};
+
+// TODO: use Gorilla as a fallback to remove GORILLA_MAXIMUM_LENGTH.
+/// Maximum number of data points that models of type Gorilla can represent per
+/// compressed segment. As models of type Gorilla use lossless compression they
+/// will never exceed the user-defined error bounds.
+const GORILLA_MAXIMUM_LENGTH: usize = 50;
+
+/// Compress the regular `uncompressed_timestamps` using a start time, end time,
+/// and a sampling interval, and `uncompressed_values` within `error_bound`
+/// using the model types in `models`. Returns `CompressionError` if
+/// `uncompressed_timestamps` and `uncompressed_values` have different lengths,
+/// otherwise the resulting compressed segments are returned as a `RecordBatch`
+/// with the schema provided by `get_compressed_segment_schema_ref()`.
+pub fn try_compress(
+    uncompressed_timestamps: &TimestampArray,
+    uncompressed_values: &ValueArray,
+    error_bound: ErrorBound,
+) -> Result<RecordBatch, ModelarDBError> {
+    // The uncompressed data must be passed as arrays instead of a RecordBatch
+    // as a TimestampArray and a ValueArray is the only supported input.
+    // However, as a result it is necessary to verify they have the same length.
+    if uncompressed_timestamps.len() != uncompressed_values.len() {
+        return Err(ModelarDBError::CompressionError(
+            "Uncompressed timestamps and uncompressed values have different lengths.".to_owned(),
+        ));
+    }
+
+    // Enough memory for end_index elements are allocated to never require
+    // reallocation as one model is created per data point in the worst cast.
+    let end_index = uncompressed_timestamps.len();
+    let mut compressed_record_batch_builder = CompressedRecordBatchBuilder::new(end_index);
+
+    // Compress the uncompressed timestamps and uncompressed values.
+    let mut current_index = 0;
+    while current_index < end_index {
+        // Create a compressed segment that represents the data points from
+        // current_index to index within error_bound where index <= end_index.
+        let compressed_segment_builder = create_next_compressed_segment(
+            current_index,
+            end_index,
+            &uncompressed_timestamps,
+            &uncompressed_values,
+            error_bound,
+        );
+        current_index += compressed_segment_builder.finish(&mut compressed_record_batch_builder);
+    }
+    Ok(compressed_record_batch_builder.finnish())
+}
+
+/// Create a compressed segment that represents the regular timestamps in
+/// `uncompressed_timestamps` and the values in `uncompressed_values` from
+/// `start_index` to index within `error_bound` where index <= `end_index`.
+fn create_next_compressed_segment<'a>(
```

Comment:
Moved the code from `create_next_compressed_segment()` into `CompressedSegmentBuilder::new()`. This was chosen instead of having both `CompressedSegmentBuilder::new()` and `CompressedSegmentBuilder::build()` as it prevents bugs where `CompressedSegmentBuilder::new()` and `CompressedSegmentBuilder::finish()` is called without `CompressedSegmentBuilder::build()`.